artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/impl/Page.java

```diff
@@ -120,105 +110,121 @@ public void setLiveCache(LivePageCache pageCache) {
-
-         if (byteRead == Page.START_BYTE) {
-            if (fileBuffer.readerIndex() + DataConstants.SIZE_INT < fileBuffer.capacity()) {
-               int messageSize = fileBuffer.readInt();
-               int oldPos = fileBuffer.readerIndex();
-               if (fileBuffer.readerIndex() + messageSize < fileBuffer.capacity() && fileBuffer.getByte(oldPos + messageSize) == Page.END_BYTE) {
-                  PagedMessage msg = new PagedMessageImpl(storageManager);
-                  msg.decode(fileBuffer);
-                  byte b = fileBuffer.readByte();
-                  if (b != Page.END_BYTE) {
-                     // Sanity Check: This would only happen if there is a bug on decode or any internal code, as
-                     // this
-                     // constraint was already checked
-                     throw new IllegalStateException("Internal error, it wasn't possible to locate END_BYTE " + b);
-                  }
+   private List<PagedMessage> readFromSequentialFile(StorageManager storage) throws Exception {
+      final List<PagedMessage> messages = new ArrayList<>();
+      final int fileSize = (int) file.size();
+      file.position(0);
+      final int minimumMessagePersistentSize = DataConstants.SIZE_INT + 2;
+      int position = 0;
+      ByteBuffer fileBuffer = null;
+      try {
+         while ((fileSize - position) >= minimumMessagePersistentSize) {
+            if (fileBuffer == null) {
+               fileBuffer = fileFactory.newBuffer(Math.min((fileSize - position), Env.osPageSize()));
+               fileBuffer.limit(0);
+            }
+            fileBuffer = readIntoFileBufferIfNecessary(file, fileBuffer, minimumMessagePersistentSize);
+            final byte startByte = fileBuffer.get();
+            if (startByte == Page.START_BYTE) {
+               final int encodedSize = fileBuffer.getInt();
+               fileBuffer = readIntoFileBufferIfNecessary(file, fileBuffer, encodedSize + 1);
+               final int endPosition = fileBuffer.position() + encodedSize;
+               //this check must be performed upfront decoding
+               if (fileBuffer.get(endPosition) == Page.END_BYTE) {
+                  final PagedMessageImpl msg = new PagedMessageImpl(storageManager);
+                  decodeInto(fileBuffer, encodedSize, msg);
+                  fileBuffer.position(endPosition + 1);
+                  assert fileBuffer.get(endPosition) == Page.END_BYTE : "decoding cannot change end byte";
                   msg.initMessage(storage);
                   if (logger.isTraceEnabled()) {
-                     logger.trace("Reading message " + msg + " on pageId=" + this.pageId + " for address=" + storeName);
+                     logger.tracef("Reading message %s on pageId=%d for address=%s", msg, pageId, storeName);
                   }
-                  readMessages++;
                   messages.add(msg);
+                  position += minimumMessagePersistentSize + encodedSize;
```

Comment:
I am a bit concerned about the new reading buffer logic.

You were supposed to fix the barrier into how mmap is reading, but you're introducing a new reading logic in the mix. It's a bit hard to review it here.

I have spent months testing this before the first release of paging was out. I'm just afraid about the amount of testing done into the new logic.

Reply:
I understand but the point of this PR is to maintain a similar speed of mmap without incurring into the time-to-safepoint issue; that's why the logic has changed.

The original implementation (pre-mmap) was instantiating for each file a giant direct `ByteBuffer`, impacting on the overall native memory footprint and causing a long context switch while making the kernel fill that buffer from its OS (file) page cache.
It has been solved using mmap, to avoid allocations and copies, but `MappedByteBuffer::get(byte)` could pause the JVM for long time: that's why I'm addressing both the issues at once: memory footprint and long time to safepoints (to save memory).

Reply:
> I have spent months testing this before the first release of paging was out. I'm just afraid about the amount of testing done into the new logic.

I'm testing the logic with the many tests we have on paging and I'm sure I will get the same robustness as before :+1: